### PR TITLE
Simplify artifact uniqueness

### PR DESCRIPTION
**Pull request by archseer**

### CODE REVIEW

File: artifact.go [hunk 1]
```diff
@@ -25,33 +25,7 @@ func (r *ArtifactRepo) Add(data interface{}) error {
...
-				return err
-			}
-		}
-		if err := tx.Commit(); err != nil {
-			return err
-		}
-	} else if err != nil {
-		return err
-	}
-	a.ID = cleanUUID(a.ID)
```
Comment (titanous) [inline artifact.go:53]: Don't delete this line.

**Comment by titanous** [comment]
Looks like a couple imports need to be dropped (see test result).

**Comment by archseer** [comment]
Resolved.

**Comment by titanous** [comment]
The master build has been fixed, so any CI errors from here on out (including the current one) should be accurate.

**Comment by archseer** [comment]
@titanous Right, so I dug into this, and figured out that we're getting errors now because of a hidden bug in the old re-creation code. Most of our tests create a blank artifact, where `URI` and `Type` fields are set to blank, and therefore not unique. But, the old method silently failed on such cases: https://github.com/flynn/flynn-controller/blob/179b2ac95d0074935cb3fb54338db02f0f1d7297/artifact.go#L29

We would get a `unique_violation`, then we'd get the old object. But, because it did not include a `deleted_at` field, it would silently fail, and all consecutive calls would then return the old object. Which, I guess, proves that you were right when you said that the old `try/catch` way of handling it was bad.

I'll resolve this by generating URIs in the createTestRelease method, but how would you like me to handle cases like https://github.com/flynn/flynn-controller/blob/master/scheduler/scheduler_test.go#L213? We use the same `foo/bar` URI in a few tests. I can either change it, or create those objects on initialisation, for all tests?

**Comment by titanous** [comment]
Hmm, previously it wasn't silently failing, it was returning the existing artifact. Maybe we should keep that behavior?

**Comment by titanous** [comment]
I should point out that since artifacts are immutable, if there is a `unique_violation`, the artifact is exactly the same.

**Comment by archseer** [comment]
Ah, it was done so by design?

**Comment by titanous** [comment]
Yeah, due to the immutability it made sense to do this instead of erroring with data that is exactly the same.

**Comment by archseer** [comment]
I see. I can just make it function the same way, by handling `unique_violation` like we did before.

**Comment by titanous** [comment]
Yeah, let's do that.

**Comment by titanous** [comment]
Thanks!

**Pull request merged**